# Moderation

* Identify inappropiate content


Traditionaly, 
* Moderators flag content by-hand
    * Time consuming
* Keyword pattern matching
    * Lacks nuance and understanding of context


## Violation Categories

* Identify violations of terms of use
* Differenciate violation type by category
    * Violance
    * Hate speech


In [1]:
from openai import OpenAI
from dotenv import load_dotenv

load_dotenv()
client = OpenAI()

## Creating a Moderations Request

In [2]:
response = client.moderations.create(
    input = "I could kill for a hamburguer."
)

### Interpreting the Results

* Categories
    * true/false indicator of category violation
* Category_scores
    * Confidence of a violation
* Flagged
    * true/false indicator of a violation

In [3]:
print(response.model_dump())

{'id': 'modr-3846', 'model': 'omni-moderation-latest', 'results': [{'categories': {'harassment': False, 'harassment_threatening': False, 'hate': False, 'hate_threatening': False, 'illicit': False, 'illicit_violent': False, 'self_harm': False, 'self_harm_instructions': False, 'self_harm_intent': False, 'sexual': False, 'sexual_minors': False, 'violence': True, 'violence_graphic': False, 'harassment/threatening': False, 'hate/threatening': False, 'illicit/violent': False, 'self-harm/intent': False, 'self-harm/instructions': False, 'self-harm': False, 'sexual/minors': False, 'violence/graphic': False}, 'category_applied_input_types': {'harassment': ['text'], 'harassment_threatening': ['text'], 'hate': ['text'], 'hate_threatening': ['text'], 'illicit': ['text'], 'illicit_violent': ['text'], 'self_harm': ['text'], 'self_harm_instructions': ['text'], 'self_harm_intent': ['text'], 'sexual': ['text'], 'sexual_minors': ['text'], 'violence': ['text'], 'violence_graphic': ['text'], 'harassment/th

```
{
    "id": "modr-3846",
    "model": "omni-moderation-latest",
    "results": [
        {
            "categories": {
                "harassment": false,
                "harassment_threatening": false,
                "hate": false,
                "hate_threatening": false,
                "illicit": false,
                "illicit_violent": false,
                "self_harm": false,
                "self_harm_instructions": false,
                "self_harm_intent": false,
                "sexual": false,
                "sexual_minors": false,
                "violence": true,
                "violence_graphic": false,
                "harassment/threatening": false,
                "hate/threatening": false,
                "illicit/violent": false,
                "self-harm/intent": false,
                "self-harm/instructions": false,
                "self-harm": false,
                "sexual/minors": false,
                "violence/graphic": false
            },
            "category_applied_input_types": {
                "harassment": ["text"],
                "harassment_threatening": ["text"],
                "hate": ["text"],
                "hate_threatening": ["text"],
                "illicit": ["text"],
                "illicit_violent": ["text"],
                "self_harm": ["text"],
                "self_harm_instructions": ["text"],
                "self_harm_intent": ["text"],
                "sexual": ["text"],
                "sexual_minors": ["text"],
                "violence": ["text"],
                "violence_graphic": ["text"],
                "harassment/threatening": ["text"],
                "hate/threatening": ["text"],
                "illicit/violent": ["text"],
                "self-harm/intent": ["text"],
                "self-harm/instructions": ["text"],
                "self-harm": ["text"],
                "sexual/minors": ["text"],
                "violence/graphic": ["text"]
            },
            "category_scores": {
                "harassment": 0.006374553647673357,
                "harassment_threatening": 0.006006826329429705,
                "hate": 9.181900965849468e-05,
                "hate_threatening": 3.9204178923762816e-05,
                "illicit": 0.01949794628932797,
                "illicit_violent": 0.0022092939645521554,
                "self_harm": 0.04096939286984131,
                "self_harm_instructions": 0.00025931199089656545,
                "self_harm_intent": 0.008569800490896309,
                "sexual": 0.00010987310729246431,
                "sexual_minors": 4.006369222920725e-06,
                "violence": 0.3503121813956502,
                "violence_graphic": 0.002388579128368805,
                "harassment/threatening": 0.006006826329429705,
                "hate/threatening": 3.9204178923762816e-05,
                "illicit/violent": 0.0022092939645521554,
                "self-harm/intent": 0.008569800490896309,
                "self-harm/instructions": 0.00025931199089656545,
                "self-harm": 0.04096939286984131,
                "sexual/minors": 4.006369222920725e-06,
                "violence/graphic": 0.002388579128368805
            },
            "flagged": true
        }
    ]
}

```

### Interpreting the Category Scores


In [ ]:
print(response.results[0].category_scores)

CategoryScores(harassment=0.006374553647673357, harassment_threatening=0.006006826329429705, hate=9.181900965849468e-05, hate_threatening=3.9204178923762816e-05, illicit=0.01949794628932797, illicit_violent=0.0022092939645521554, self_harm=0.04096939286984131, self_harm_instructions=0.00025931199089656545, self_harm_intent=0.008569800490896309, sexual=0.00010987310729246431, sexual_minors=4.006369222920725e-06, violence=0.3503121813956502, violence_graphic=0.002388579128368805, harassment/threatening=0.006006826329429705, hate/threatening=3.9204178923762816e-05, illicit/violent=0.0022092939645521554, self-harm/intent=0.008569800490896309, self-harm/instructions=0.00025931199089656545, self-harm=0.04096939286984131, sexual/minors=4.006369222920725e-06, violence/graphic=0.002388579128368805)


```
CategoryScores(
    harassment=0.006374553647673357, 
    harassment_threatening=0.006006826329429705, 
    hate=9.181900965849468e-05, 
    hate_threatening=3.9204178923762816e-05, 
    illicit=0.01949794628932797, 
    illicit_violent=0.0022092939645521554, 
    self_harm=0.04096939286984131, 
    self_harm_instructions=0.00025931199089656545, 
    self_harm_intent=0.008569800490896309, 
    sexual=0.00010987310729246431, 
    sexual_minors=4.006369222920725e-06, 
    violence=0.3503121813956502, 
    violence_graphic=0.002388579128368805, 
    harassment/threatening=0.006006826329429705, 
    hate/threatening=3.9204178923762816e-05, 
    illicit/violent=0.0022092939645521554, 
    self-harm/intent=0.008569800490896309, 
    self-harm/instructions=0.00025931199089656545, 
    self-harm=0.04096939286984131, 
    sexual/minors=4.006369222920725e-06, 
    violence/graphic=0.002388579128368805)
```

> Violence = 0.35

* Large numbers -> greater certainty of violation
* Numbers $\neq$ probabilities

### Considerations for implementing moderation

* Tune thresholds for each use case
* Stricter thresholds may result in fewer false negatives
* More lenient thresholds may result in fewer false positives